buildtest/buildsystem/base.py

```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+
+        # every test starts with shebang line
+        lines = [self.shebang]
+
+        # if 'module' defined in Buildspec add modules to test
+        if self.recipe.get("module"):
+            for module in self.recipe.get("module"):
+                lines.append(module)
+
+        # add compile command to test
+        lines.append(" ".join(cmd))
+
+        # build run command that launches generated executable. Format: <exec> <args>
+        # if 'exec_args' is not defined it will return 'None' which means no arguments to be passed when running executable
+        run = [
+            self.executable,
+            self.compiler_recipe[self.compiler_name].get("exec_args"),
+        ]
+
+        # remove 'None' in case 'exec_args' is not defined in Buildspec
+        run = list(filter(None, run))
+
+        # add run line to test
+        lines.append(" ".join(run))
+        return lines
+
+    def detect_lang(self, sourcefile):
+        """This method will return the Programming Language based by looking up  file extension of source file."""
+
+        ext = os.path.splitext(sourcefile)[1]
+
+        # if ext not in self.lang_ext_table then raise an error. This table consist of all file extensions that map to a Programming Language
+        if ext not in self.lang_ext_table:
+            raise BuildTestError(
+                f"Unable to detect Program Language based on extension: {ext} in file {sourcefile}"
+            )
+        # Set Programming Language based on ext. Programming Language could be (C, C++, Fortran)
+        lang = self.lang_ext_table[ext]
+        return lang
+
+    def generate_compile_cmd(self):
+        """This method generates the compilation line and returns the output as a list. The compilation line depends
+           on the the language detected that is stored in variable ``self.lang``.
+        """
+        cmd = []
+
+        # Generate C compilation line
+        if self.lang == "C":
+            cmd = [
+                self.cc,
+                self.cppflags,
+                self.cflags,
+                "-o",
+                self.executable,
+                self.source,
+                self.ldflags,
+            ]
+
+        # Generate C++ compilation line
+        elif self.lang == "C++":
+            cmd = [
+                self.cxx,
+                self.cppflags,
+                self.cxxflags,
+                "-o",
+                self.executable,
+                self.source,
+                self.ldflags,
+            ]
+
+        # Generate Fortran compilation line
+        elif self.lang == "Fortran":
+            cmd = [
+                self.fc,
+                self.cppflags,
+                self.fflags,
+                "-o",
+                self.executable,
+                self.source,
+                self.ldflags,
+            ]
+        # remove any None from list
+        return list(filter(None, cmd))
+
+
+class GNUCompiler(CompilerBuilder):
```

Comment:
This is a great start! This is exactly what I had in mind - see if you can simplify the CompilerBuilder base so that it doesn't have all that if else elif logic and you can just put shared functions there, and specifics in each of the classes here.